juju_spell/commands/list_models.py

```
@@ -13,17 +17,51 @@ async def execute(
+            outputs = load_cache_data(kwargs["controller_config"].uuid, self.logger)
+
+        if kwargs["refresh"] or isinstance(outputs, JujuSpellError):
             models = list(
                 await self.get_filtered_model_names(
                     controller=controller,
                     models=None,
                     model_mappings=kwargs["controller_config"].model_mapping,
                 )
             )
-        else:
-            models = ["TODO: read from cache; not implemented."]
+            outputs = save_cache_data(controller, models, kwargs["refresh"], self.logger)
 
-        outputs = {"refresh": kwargs["refresh"], "models": models}
         self.logger.debug("%s list models: %s", controller.controller_uuid, outputs["models"])
         return outputs
+
+
+def save_cache_data(
+    controller: Controller, models: List[str], refresh: bool, logger: Logger
+) -> Dict[str, Any]:
+    """Gracefully save cache data to default cache directory."""
+    cache_data = {
+        "uuid": controller.controller_uuid,
+        "name": controller.controller_name,
+        "models": models,
+        "refresh": refresh,
+        "timestamp": time(),
+    }
+    error_message_template = "%s list models failed to save cache: %s."
+    try:
+        save_to_cache(cache_data, controller.controller_uuid)
+    except JujuSpellError as error:
+        logger.debug(error_message_template, controller.controller_uuid, str(error))
+    return cache_data
+
+
+def load_cache_data(uuid: str, logger: Logger) -> Union[JujuSpellError, Dict[str, Any]]:
+    """Gracefully load cache data from default cache directory."""
+    error_message_template = "%s list models failed to load cache: %s."
+    try:
+        cache_data = load_from_cache(uuid)
+        cache_data["refresh"] = False
+    except JujuSpellError as error:
+        logger.debug(error_message_template, uuid, str(error))
+        return JujuSpellError(error_message_template.format(uuid, str(error)))
```

I think that this should raise an error or return None if cache does not exists.

---

ACK